Start autogenerating WebExtensionAPIWebNavigationEvent.idl and add the necessary files so everything compiles

b-weinstein opened the pull request:
#### f485872bdb43055326125c16f9ac0dd62e421dc1
<pre>
Start autogenerating WebExtensionAPIWebNavigationEvent.idl and add the necessary files so everything compiles
<a href="https://bugs.webkit.org/show_bug.cgi?id=248845">https://bugs.webkit.org/show_bug.cgi?id=248845</a>
rdar://102820594

Reviewed by Timothy Hatcher.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm: Added.
(WebKit::WebExtensionAPIWebNavigationEvent::invokeListenersWithArgument): Invoke the listener with a single argument (which will be after checking the URL against the filter).
(WebKit::WebExtensionAPIWebNavigationEvent::addListener): Add the listener.
(WebKit::WebExtensionAPIWebNavigationEvent::removeListener): Remove the listener.
(WebKit::WebExtensionAPIWebNavigationEvent::hasListener): Determine if the listener is installed.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebNavigationEvent.h: Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebNavigationEvent.idl.
(WebKit::WebExtensionAPIWebNavigationEvent::listeners const):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebNavigationEvent.idl: Use [NeedsPage] on the two methods that need the page.

Canonical link: <a href="https://commits.webkit.org/257501@main">https://commits.webkit.org/257501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/437ed0dcb39254c294562df52407ec4f1f657712

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99178 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8387 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/32312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108575 "Built successfully") | 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/8940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106510 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104914 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/8940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/32312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/91685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/8940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/32312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/2272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/32312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2164 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5155 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/7216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/32312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3640 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
```diff
@@ -0,0 +1,90 @@
+void WebExtensionAPIWebNavigationEvent::addListener(WebPage* page, RefPtr<WebExtensionCallbackHandler> listener, NSDictionary *filter, NSString **exceptionString)
+{
+    // FIXME: Save the filter associated with each of these listeners.
+
```
Should we add a call here to `hasListener()` and return early here if the listener already exists?

xeenon replied:
I think extensions can add the same listener multiple times. What do we do in Safari currently?

kiaraarose replied:
Looks like we don't allow the same listener more than once. Apparently "In Chrome, adding a callback function more than once is a no-op."

b-weinstein replied:
I was just matching what was in WebExtensionAPIEventCocoa.mm.

I think you're right, we should do that in both places.

b-weinstein replied:
I'll handle this in a follow-up, make sure we get it right.

b-weinstein committed:
Start autogenerating WebExtensionAPIWebNavigationEvent.idl and add the necessary files so everything compiles
https://bugs.webkit.org/show_bug.cgi?id=248845
rdar://102820594

Reviewed by Timothy Hatcher.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm: Added.
(WebKit::WebExtensionAPIWebNavigationEvent::invokeListenersWithArgument): Invoke the listener with a single argument (which will be after checking the URL against the filter).
(WebKit::WebExtensionAPIWebNavigationEvent::addListener): Add the listener.
(WebKit::WebExtensionAPIWebNavigationEvent::removeListener): Remove the listener.
(WebKit::WebExtensionAPIWebNavigationEvent::hasListener): Determine if the listener is installed.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebNavigationEvent.h: Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebNavigationEvent.idl.
(WebKit::WebExtensionAPIWebNavigationEvent::listeners const):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebNavigationEvent.idl: Use [NeedsPage] on the two methods that need the page.

Canonical link: https://commits.webkit.org/257501@main
